apps/balancer-tools/src/app/apr/(utils)/tokenYield.ts

```diff
@@ -20,42 +20,122 @@ export const publicClient = createPublicClient({
+          address: rateProviderAddress,
+          token: { symbol, address: tokenAddress },
+        }) => ({
+          address: tokenAddress,
+          symbol,
+          yield: await getAPRFromRateProviderInterval(
+            rateProviderAddress,
+            now - SECONDS_IN_DAY,
+            now,
+          ),
+        }),
+      ),
+  );
+}
+
+async function getAPRFromRateProviderInterval(
+  rateProviderAddress: `0x${string}`,
+  timeStart: number,
+  timeEnd: number,
+) {
+  const { endRate, startRate } = await getIntervalRates(
+    rateProviderAddress,
+    timeStart,
+    timeEnd,
+  );
+
+  return getAPRFromRate(startRate, endRate, timeStart, timeEnd);
+}
+
+function getAPRFromRate(
+  rateStart: number,
+  rateEnd: number,
+  timeStart: number,
+  timeEnd: number,
+) {
+  const duration = timeEnd - timeStart;
+  const rateOfReturn = (rateEnd - rateStart) / rateStart;
+  const annualScalingFactor = SECONDS_IN_YEAR / duration;
+
+  const APR = rateOfReturn * annualScalingFactor * 100;
+
+  return APR;
+}
+
+async function getPoolTokensRateProviders(
+  chain: string,
+  poolId: `0x${string}`,
+) {
+  const data = await pools.gql(String(chain)).PoolRateProviders({ poolId });
+  if (!data.pool?.priceRateProviders?.length)
+    return [
+      {
+        address: "0x1a8F81c256aee9C640e14bB0453ce247ea0DFE6F",
+        token: {
+          address: "0xae78736cd615f374d3085123a210448e74fc6393",
+          symbol: "rETH",
+        },
+      },
+    ];
+
+  return data.pool?.priceRateProviders;
+}
+
+async function getIntervalRates(
+  rateProviderAddress: `0x${string}`,
+  timeStart: number,
+  timeEnd: number,
+) {
   const data = await blocks.gql(String(NetworkChainId.ETHEREUM)).Blocks({
-    timestamp_gte: now - SECONDS_IN_DAY,
-    timestamp_lt: now,
+    timestamp_gte: timeStart,
+    timestamp_lt: timeEnd,
   });
 
-  const block24hAgo = data.blocks[0].number;
+  const blockStart = data.blocks[0].number;
 
-  const currentRate = await publicClient.readContract({
-    address: "0x1a8F81c256aee9C640e14bB0453ce247ea0DFE6F",
-    abi: rateProviderAbi,
-    functionName: "getRate",
-  });
+  const [endRate, startRate] = await Promise.all([
+    getRateAtBlock(rateProviderAddress),
+    getRateAtBlock(rateProviderAddress, blockStart),
+  ]);
+
+  return { endRate: Number(endRate), startRate: Number(startRate) };
+}
 
-  const rate24hAgo = await publicClient.readContract({
-    address: "0x1a8F81c256aee9C640e14bB0453ce247ea0DFE6F",
+async function getRateAtBlock(
+  rateProviderAddress: `0x${string}`,
+  blockNumber?: number,
+) {
+  const args = {
+    address: rateProviderAddress,
     abi: rateProviderAbi,
     functionName: "getRate",
-    blockNumber: BigInt(block24hAgo),
-  });
+    blockNumber: blockNumber ? BigInt(blockNumber) : undefined,
+  } as const;
+
+  let rate;
+  try {
+    rate = await publicClient.readContract(args);
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(e);
+    rate = -1;
```

Comment:
we should log it somewhere when this happens and just default to 0 not to mess with the other numbers